### ISSUE TITLE
Consider reworking the pattern used for the validation spans

### ISSUE THREAD

**Issue by PinpointTownes** [OrchardCMS/OrchardCore#1462]
In many `.cshtml` files, a default "hardcoded" error message is included in the validation spans. E.g:

```html
<span asp-validation-for="UserName" class="text-danger">@T["The user name is required."]</span>
```

This is very unfortunate because `ValidationMessageTagHelper` doesn't override this message, which means every other source of validation error (including errors manually added via `AddModelError`) are never rendered.

@sebastienros is there any reason this pattern was adopted? Perhaps to make localization of the "required field" messages easier?

**Comment by PinpointTownes** [comment]
Looks like this pattern is massively used in OC (including in the OpenID module :sob:)

Instead, I suggest we enable MVC's transparent Data Annotations localization and replace things like...

```csharp
private void ValidateViewModel(EditRemoteClientViewModel model)
{
    if (String.IsNullOrWhiteSpace(model.ClientName))
    {
        ModelState.AddModelError(nameof(EditRemoteClientViewModel.ClientName), S["The client name is mandatory."]);
    }

    if (String.IsNullOrWhiteSpace(model.ApiKey))
    {
        ModelState.AddModelError(nameof(EditRemoteClientViewModel.ApiKey), S["The api key is mandatory."]);
    }
}
```

... by the regular `RequiredAttribute`:

```csharp
public class EditRemoteClientViewModel
{
    public string Id { get; set; }
    [Required(ErrorMessage = "The client name is mandatory.")]
    public string ClientName { get; set; }
    [Required(ErrorMessage = "The api key is mandatory.")]
    public string ApiKey { get; set; }
}
```

This way, we should be able to remove the hardcoded errors from the view, without loosing the ability to localize the validation messages defined via Data Annotations.

**Comment by sebastienros** [comment]
It's definitely probably that ASP.NET Core lets us use `IStringLocalizer` for the data annotations. It was not the case before. But we'll still need custom validation in the controllers, so I'd be happy to fix any issue you found with the way we did it. This must be a local issue as we use it everywhere and validation has always worked just fine.

**Comment by PinpointTownes** [comment]
@sebastienros clone https://github.com/OrchardCMS/OrchardCore/pull/1463/files locally. Enable the OpenID module, create a new app. Then, create another app with the same client identifier. On my machine, I'm unable to see the error message indicating the client identifier already exists.

Concerning the localization of data annotations, I believe it's been implemented in 1.0 (but non-validation attributes are only localized since 1.1).

**Comment by PinpointTownes** [comment]
> This must be a local issue as we use it everywhere and validation has always worked just fine.

Not sure when you tried it last time, but with https://github.com/OrchardCMS/OrchardCore/issues/1461, I really doubt we can say it "works just fine" :sweat_smile:

**Comment by sebastienros** [comment]
> This must be a local issue

**Comment by sebastienros** [comment]
But we'll find a solution

**Comment by PinpointTownes** [comment]
How can that be a local issue? :sweat_smile: 

`ValidationMessageTagHelper` deliberately doesn't override the inner content of a `span` block that already contains some data: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.TagHelpers/ValidationMessageTagHelper.cs#L93-L104

Yet, since many views provide a "default error", the validation tag helper doesn't work as you'd expect. I personally call that... a design flaw :smile:

**Comment by PinpointTownes** [comment]
@sebastienros what's the final call for this issue? Can we at least remove the hardcoded messages in the spans?

**Comment by sebastienros** [comment]
I think I understand the issue now. Your problem is that if we use an `asp-validation-for` tag with a custom text inside, then there is no way from the controller to define a different message.

What is your recommended design? That we still add a `span` with this attribute, but without any text, such that the model error can be injected from a data annotation or a custom call to `AddModelError` ?

**Comment by PinpointTownes** [comment]
> That we still add a span with this attribute, but without any text, such that the model error can be injected from a data annotation or a custom call to AddModelError ?

Yup.

**Comment by sebastienros** [comment]
Good, let's do this. This issue will track the progress.

**Comment by PinpointTownes** [comment]
What do we do for the error messages? Do we convert them as `[Required(ErrorMessage = "localized error message")]` or should we use the default MVC error messages? (enabled via `[Required]`)

**Comment by hishamco** [comment]
@Skrypt is this done in the branch that you mentioned?

**Comment by Skrypt** [comment]
Yes it works but the pattern still needs to be changed in some views to be more consistent.

**Comment by hishamco** [comment]
I will check your PR as well as what are the changes required in some views to be more consistent

**Comment by kizaski** [comment]
This seems partially fixed by [PR #8213](https://github.com/OrchardCMS/OrchardCore/pull/8213). As in, there are no more hard coded errors in any cshtml, but there is still no use of `RequiredAttribute`.

Should the related code still be refactored to use `RequiredAttribute`?

**Comment by gvkries** [comment]
There are still hard coded error messages in some validation tags, but nothing critical IMHO. Same goes for the `RequiredAttribute`, I don't think this is necessary to change.
